### ISSUE TITLE
Empty / NonEmpty Filters apply to dateFieldFilter only. 

### ISSUE THREAD

**Issue by hfcjweinstock** [EncompassRest/EncompassRest#83]
We need to redefine Empty/nonEmpty Filters to be a subclass of DateFieldFilter.  the value is required but ignored on  EM's side.

c/p from EM documentation:
Sample Payloads
Return all data where Loan.LastModified is null. In this case, the value doesn't matter. Applies to dates only.

Sample Request
 {
  "canonicalName": "Loan.LastModified",
  "value": "2015-05-23T21:05:46Z",
  "matchType": "isEmpty"
}
Return all data where Loan.LastModified is not null. In this case, the value also doesn't matter. Applies to dates only.

Sample Request
 {
  "canonicalName": "Loan.LastModified",
  "value": "2015-05-23T21:05:46Z",
  "matchType": "isNotEmpty"
}

**Comment by TylerBrinkley** [comment]
`EmptyFieldFilter` and `NotEmptyFieldFilter` apply to all fields not just date fields. This seems like an Encompass bug to me.

**Comment by hfcjweinstock** [comment]
nope.  only date fields have an "Empty"  everything else in query land has a value

**Comment by hfcjweinstock** [comment]
and it is documented.....

**Comment by TylerBrinkley** [comment]
I'm not seeing anywhere in the documentation that a `matchType` of `isEmpty` and `isNotEmpty` should only apply to date fields.

**Comment by hfcjweinstock** [comment]
i copied and pasted from the dev connect.  in the sample payloads.  like i said.  documented, but poorly.

**Comment by hfcjweinstock** [comment]
use the existing report system as a guide. in the reports tab of encompass, there is only an "isEmpty/isNotEmpty" option for date fields.

**Comment by TylerBrinkley** [comment]
The sample payloads just show them using the `isEmpty` and `isNotEmpty` `matchType`s for a date field. It does not show that they're not supported for string or numeric fields.

**Comment by hfcjweinstock** [comment]
Return all data where Loan.LastModified is null. In this case, the value doesn't matter. **Applies to dates only.**

**Comment by TylerBrinkley** [comment]
Ugh. What if we want to filter on an empty numeric field and not include 0 values. I know this isn't currently possible in the SDK but it really should. Maybe I'll post to the forums about it.

**Comment by hfcjweinstock** [comment]
you theoretically cannot have an empty numeric value, but using the existing query language, you use a stringQuery and look for blank,

**Comment by TylerBrinkley** [comment]
Does string query work on non string fields?

**Comment by hfcjweinstock** [comment]
probably

**Comment by TylerBrinkley** [comment]
testing sdk right now

**Comment by hfcjweinstock** [comment]
sdk does.  i know 100%.  does the api....

**Comment by TylerBrinkley** [comment]
I'm getting an exception in the SDK of "Error converting data type varchar to numeric." when using a numeric field in a `StringFieldCriterion`.

**Comment by hfcjweinstock** [comment]
further verification on this issue.  if you put in anything but a valid datetime into the value while using the isEmpty, it rejects the request.  solution options : leave the classes how they are but add a value that returnes datetime.minval OR change the class names to be empty/nonempty DateFieldFilter.(i thought about putting it under the date field filter, but that change datefield to nonsealed.  not sure if that is an issue or not.  plus precision is not needed.

**Comment by hfcjweinstock** [comment]
hmm...ok...that used to work......

**Comment by TylerBrinkley** [comment]
Nevermind, I was using string.Empty instead of null.

**Comment by TylerBrinkley** [comment]
I'm not getting any results though I know some fields are empty.

**Comment by hfcjweinstock** [comment]
that is because when you query the database, the field is a numeric field.  there are no empty numeric fields. not allowed. look at those fields in the pipeline screen you will see 0.

**Comment by TylerBrinkley** [comment]
No, some show up in the pipeline as blank.

**Comment by hfcjweinstock** [comment]
in 162,000 files, we have 2 loans with a blank loan amount, and im betting it is a glitch.

**Comment by hfcjweinstock** [comment]
you cannot query for blank values in the pipeline view.  it is a glitch.  submit a bug!

**Comment by hfcjweinstock** [comment]
for money values at least.

**Comment by hfcjweinstock** [comment]
support ticket has been created for this issue with Ellie Mae.
(case # 02149812)

**Comment by TylerBrinkley** [comment]
Nice, is that for the SDK or API's? Also what did you request?

**Comment by hfcjweinstock** [comment]
it is a system level bug.  SDK/API/Smart Client.  i let them know that if the numeric field is blank in the reporting database, any criteria imposed on that field will exclude the loan entirely.

**Comment by TylerBrinkley** [comment]
It will include the loan if you specify 0 though.

**Comment by hfcjweinstock** [comment]
not on my system

**Comment by TylerBrinkley** [comment]
The following retrieved loans with blank values for the loan term field.

```c#
using (var session = new Session())
{
    session.Start("serverUri", "userId", "password");
    const string numericField = "Fields.4";
    using (var cursor = session.Reports.OpenReportCursor(new StringList(new[] { numericField }), new NumericFieldCriterion(numericField, 0, OrdinalFieldMatchType.Equals)))
    {
        foreach (var data in cursor.GetItems(0, cursor.Count).Cast<LoanReportData>())
        {
            var fieldValue = data[numericField];
            Console.WriteLine($"{data.Guid}: {(fieldValue == null ? "null" : $"'{fieldValue}'")}");
        }
    }
}
```

**Comment by hfcjweinstock** [comment]
ok, but my point is you are not searching for null, you are searching for 0 and looking for results that are null.

**Comment by TylerBrinkley** [comment]
I think that's by design, unfortunately.

**Comment by TylerBrinkley** [comment]
Also, the value's I got back were always automatically converted to 0 and did not return as null.

**Comment by hfcjweinstock** [comment]
either way, the isEmpty is a date only match type.  how should it be implemented?

**Comment by TylerBrinkley** [comment]
I think we should talk to Sai about it and see if it's possible it will be implemented for string or numeric fields. If so then we can keep it how it is and for now send a `DateTime.MinValue` value property. If not then we probably ought to rename them to `EmptyDateFieldFilter` and `NotEmptyDateFieldFilter`.

**Comment by TylerBrinkley** [comment]
It appears Ellie Mae has updated the endpoints to support the `isEmpty` and `isNotEmpty` `matchType`s for string and numeric types. Also the `value` is no longer required. #123 adds tests to confirm this.

**Comment by TylerBrinkley** [comment]
Reopening due to string types not currently supported by Ellie Mae.

**Comment by hfcjweinstock** [comment]
is Empty Date filter does not work without value on create cursor. numeric works fine with a value of a DateTime in the EmpyField filter.

**Comment by TylerBrinkley** [comment]
The `value` property is no longer required.

**Comment by hfcjweinstock** [comment]
try creating a cursor, not viewpipeline

**Comment by hfcjweinstock** [comment]
pull #135 corrects the issue for use with pipeline cursor, but issue is still open for string field values.
